src/Shipping.php

```diff
@@ -0,0 +1,253 @@
+	 * @param WC_Product $product           Product for which we want to generate the shipping column.
+	 * @return array|null                   Name and cost for the best found rate or null in case nothing was found.
+	 */
+	private static function get_best_shipping_with_cost( $shipping_location, $shipping_methods, $product ) {
+
+		// Since in a shipping zone all locations are treated the same we will perform the calculations for the first one.
+		$package = self::put_product_into_a_shipping_package( $product, $shipping_location );
+		$rates   = array();
+
+		// Substitute default customer location and billing for.
+		WC()->customer = new \WC_Customer( get_current_user_id() );
+		WC()->customer->set_billing_location( $shipping_location['country'], $shipping_location['state'] );
+		WC()->customer->set_shipping_location( $shipping_location['country'], $shipping_location['state'] );
+
+		// By using the filter we can trick the get_rates_for_package to continue calculations even without having the Cart defined.
+		add_filter( 'woocommerce_shipping_free_shipping_is_available', array( static::class, 'is_free_shipping_available' ), 10, 3 );
+		foreach ( $shipping_methods as $shipping_method ) {
+			$rates += $shipping_method->get_rates_for_package( $package );
+		}
+		remove_filter( 'woocommerce_shipping_free_shipping_is_available', array( static::class, 'is_free_shipping_available' ), 10 );
+
+		// Check if shipping methods have returned any valid rates.
+		if ( empty( $rates ) ) {
+			return null;
+		}
+
+		$best_rate = self::calculate_best_rate( $rates );
+		return $best_rate;
+	}
+
+	/**
+	 * Pick the best rate from an array of rates.
+	 *
+	 * @since x.x.x
+	 *
+	 * @param  array $rates List of shipping rates.
+	 * @return array        Name and cost for the best found rate or null in case nothing was found.
+	 */
+	private static function calculate_best_rate( $rates ) {
+		$best_cost = INF;
+		$best_name = '';
+		foreach ( $rates as $rate ) {
+			$shipping_cost = (float) $rate->get_cost();
+			$shipping_tax  = (float) $rate->get_shipping_tax();
+			$total_cost    = $shipping_cost + $shipping_tax;
+			if ( $total_cost < $best_cost ) {
+				$best_cost = $total_cost;
+				$best_name = $rate->get_label();
+			}
+		}
+
+		return array(
+			'cost' => wc_format_decimal( $best_cost, 2 ),
+			'name' => $best_name,
+		);
+	}
+
+	/**
+	 * Helper function that packs products into a package structure required by the shipping methods.
+	 *
+	 * @since x.x.x
+	 *
+	 * @param WC_Product $product  Product to package.
+	 * @param array      $location Product destination location.
+	 * @return array               Product packed into a package for use by shipping methods.
+	 */
+	public static function put_product_into_a_shipping_package( $product, $location ) {
+		$cart_item = array(
+			'key'          => 0,
+			'product_id'   => $product->get_id(),
+			'variation_id' => null,
+			'variation'    => null,
```

Comment:
❔ 

Just curious that the variation_id is null but we are still getting the correct shipping if the variation has a shipping class?

Reply:
The shipping classes are only relevant for the Flat Rate Shipping. In that method the `find_shipping_classes` is used for finding classes. Inside we have the `get_shipping_class` being called on the `$product` that is under the `data` key in the `$cart_item`. The `variation_id` plays no role in this process - it is used elsewhere. So as long as we have correct item under `data` we are OK.